add MaxNetworkRetries parameter to backend configuration

Compaurum opened the pull request:
implement mechanism for retrying failure request caused by network errors.
Solves problem described here https://github.com/stripe/stripe-go/issues/528 and here https://github.com/stripe/stripe-go/issues/529

brandur-stripe commented on stripe.go:
```diff
@@ -283,6 +296,8 @@ func (s *BackendConfiguration) NewRequest(method, path, key, contentType string,
 			}
 
 			req.Header.Add("Idempotency-Key", idempotency)
+		} else if strings.ToUpper(req.Method) == http.MethodPost || strings.ToUpper(req.Method) == http.MethodPut || strings.ToUpper(req.Method) == http.MethodDelete {
```
Given that the use of `PUT` is somewhat speculative already, I wonder if we should just add `PATCH` in here as well for good measure.

Compaurum replied:
@brandur-stripe Add `PATCH` or replace `PUT` with `PATCH`?

brandur-stripe replied:
Add `PATCH`.

brandur-stripe commented on stripe.go:
```diff
@@ -315,7 +330,24 @@ func (s *BackendConfiguration) Do(req *http.Request, v interface{}) error {
+		res, err = s.HTTPClient.Do(req)
+		if s.shouldRetry(err, res, retry) {
+			sleep := s.sleepTime(retry)
+			if LogLevel > 2 {
```
I'm kind of thinking that maybe this should be `LogLevel > 0`. Errors that occur here are still errors and should probably appear in the logs.

Compaurum replied:
ok

brandur-stripe commented on stripe.go:
```diff
@@ -485,3 +517,44 @@ func initUserAgent() {
+	// number of num_retries so far as inputs.
+	delay = minNetworkRetryDelay + minNetworkRetryDelay*numRetries*numRetries
+
+	//Do not allow the number to exceed maxNetworkRetryDelay.
```
Mind just editing this to have a space after the `//`?

brandur-stripe commented on stripe.go:
```diff
@@ -485,3 +517,44 @@ func initUserAgent() {
+		delay = maxNetworkRetryDelay
+	}
+
+	//Apply some jitter by randomizing the value in the range of 75%-100%.
```
And same here.

brandur-stripe commented on stripe.go:
```diff
@@ -44,6 +45,11 @@ const TotalBackends = 2
+// maxNetworkRetryDelay and minNetworkRetryDelay defines sleep time in milliseconds between
+// tries to send HTTP request again after network failure.
+const maxNetworkRetryDelay = 5000
+const minNetworkRetryDelay = 500
```
So I think just for maximum correctness and because Go supports this natively, we should use `time.Duration` for both of these. So something like `5000 * time.Millisecond`.

brandur-stripe commented on stripe.go:
```diff
@@ -72,13 +78,15 @@ func (a *AppInfo) formatUserAgent() string {
 type Backend interface {
 	Call(method, path, key string, body *form.Values, params *Params, v interface{}) error
 	CallMultipart(method, path, key, boundary string, body io.Reader, params *Params, v interface{}) error
+	SetMaxNetworkRetry(maxNetworkRetry int)
```
It seems that using "retries" instead of "retry" would make this more fluid, and also be consistent with `MaxNetworkRetries` in `BackendConfiguration` below. Any reason you went with the singular version of the word here?

brandur-stripe commented on stripe.go:
```diff
@@ -485,3 +517,44 @@ func initUserAgent() {
+func (s *BackendConfiguration) sleepTime(numRetries int) (delay int) {
+	// Apply exponential backoff with minNetworkRetryDelay on the
+	// number of num_retries so far as inputs.
+	delay = minNetworkRetryDelay + minNetworkRetryDelay*numRetries*numRetries
```
It looks like you were using stripe-ruby's implementation as a template to build this, but that you deviated in a few ways from it  — would you mind just explaining the rationale for why? It's not too big of a deal, but of course maintenance will be slightly easier if all of these are using the same algorithm.

Also, it might be somewhat nice just to make a note in here that we expect the delay in here to be measured in milliseconds and how given that, it's probably okay to calculate randomness with an integer.

Compaurum replied:
I think this logic is more clear for go programmer. Also I ruby version delay time could be bigger then maxDelay (because of jitter). I don't think that it's very good. Just say if it's not clear and I should do the identical logic as in ruby.

brandur-stripe replied:
Yeah agreed that it shouldn't go past max delay. Okay, it's fine to leave as is.

brandur-stripe commented on stripe.go:
```diff
@@ -485,3 +517,44 @@ func initUserAgent() {
+}
+
+//sleepTime calculates sleeping/delay time in milliseconds between failure and a new one request.
+func (s *BackendConfiguration) sleepTime(numRetries int) (delay int) {
```
Can you fix the spacing after the `//` here? And similar to the constants above, we should have this return a `time.Duration` instead of an integer. Lastly, is there a good reason to make this a named return value? I'd prefer to avoid those if possible because they're somewhat confusing unless you already know how they work.

brandur-stripe commented on stripe.go:
```diff
@@ -485,3 +517,44 @@ func initUserAgent() {
+	}
+
+	//Apply some jitter by randomizing the value in the range of 75%-100%.
+	jitter := rand.Intn(delay / 4)
```
Also looks like a bit of deviation from the stripe-ruby implementation. It's not the end of the world, but is there any particular reason to choose `4` here?

Compaurum replied:
In my opinion stripe-ruby value (`2`) is too big. Delay should increase smoothlyIn ruby implementation previous
 value could be greater then next. For example possible answer in ruby is:  `6939
15189
18939
26439
24439`.  So on 4-th iteration value is bigger then on 5-th.   Anyway if you want, I will change it - it's not the end of the world :)

brandur-stripe replied:
Yeah, it's not a big deal. I was just curious where the `4` in particular came from.

brandur-stripe commented:
(Just adding a comment here because we're currently seeing a false positive on our response checking script.)

brandur-stripe commented:
@Compaurum Thanks for taking a look at that feedback. Let me know when this is ready for re-review.

brandur-stripe commented on stripe.go:
```diff
@@ -250,6 +263,7 @@ func (s *BackendConfiguration) CallMultipart(method, path, key, boundary string,
 // NewRequest is used by Call to generate an http.Request. It handles encoding
 // parameters and attaching the appropriate headers.
 func (s *BackendConfiguration) NewRequest(method, path, key, contentType string, body io.Reader, params *Params) (*http.Request, error) {
+	method = strings.ToUpper(method)
```
Did you end up putting in the call to `ToUpper` for any particular reason? I just grepped the library and it seems like all verbs are specified as uppercase already, so I don't think we need this.

It also occurred to me that we should really just be using the `http` constants like `http.MethodPut` instead of directly passing verb names. I'll put together a patch that does that separately.

brandur-stripe replied:
And just to be clear, I'd have a preference for dropping `ToUpper` for now if it's not needed.

brandur-stripe replied:
I'm moving us to `net/http` constants in #587.

brandur-stripe commented on stripe.go:
```diff
@@ -283,6 +297,8 @@ func (s *BackendConfiguration) NewRequest(method, path, key, contentType string,
 			}
 
 			req.Header.Add("Idempotency-Key", idempotency)
+		} else if method == http.MethodPost || method == http.MethodPut || method == http.MethodPatch || method == http.MethodDelete {
```
Thanks for making this change!

Now that I look at this again, would you mind creating a new function called `isHTTPWriteMethod` (or such) and abstract this line into it? I think it'll be cleaner code-wise and more suggestive of why this check is being made.

Compaurum commented:
@brandur-stripe Sorry, I forget to mention you after those changes. I've added function `isHTTPWriteMethod`, removed redundant `ToUpper` and fixed some conflicts with master branch.
Waiting for your call

brandur-stripe commented:
> @brandur-stripe Sorry, I forget to mention you after those changes. I've added function isHTTPWriteMethod, removed redundant ToUpper and fixed some conflicts with master branch.

No worries! I just feel bad because we ended up leaving you hanging for so long.

Anyway, thanks for the quick changes! This looks great.

brandur-stripe commented:
Released as 35.0.0.
